### PR TITLE
Refactor InstallPlan

### PR DESCRIPTION
**Pull request by dcoutts**
Currently the `InstallPlan` structure keeps track of the state of execution of the plan. It does this by tracking the state of each package: Processing, Installed, Failed. So one way of looking at it is that the `InstallPlan` type itself is used as the state of a traversal over an `InstallPlan`.

It would be nicer to be able to traverse/execute an `InstallPlan` independently from modifying the `InstallPlan` itself, or equivalently to keep the state of the traversal of a plan separate from the plan. So that's what this patch set does.

The motivation is twofold:

 1. It is conceptually clearer to have the `InstallPlan` just be immutable, and have separate execute/traversals. It also reduces coupling, e.g. things that construct an `InstallPlan` no longer need to know what type of result or failure will be used by the code that traverses/executes the plan later.
 2. The current approach forces the install failure type to be serialisable (if the overall `InstallPlan` is to be serialisable). This means we cannot use `SomeException` as part of the failure type in the `new-build` code path. The `new-build` code needs the `InstallPlan` to be serialisable since it caches it on disk. By separating traversal from the plan type we can make the original `InstallPlan` serialisable since it never mentions anything about the failure type. In turn this will let use use `SomeException` for the failure type which will then let us do better exception handling in the `new-build` code path.

So this stuff is also a pre-requisite for some of the `new-build` exception handling improvements.

### CODE REVIEW

**Comment by ezyang** [comment]
I'm merging this with master.

File: cabal-install/Distribution/Client/InstallPlan.hs [hunk 1]
```diff
@@ -743,74 +565,354 @@ closed = null . PlanIndex.brokenPackages
...
+        (nextpkgs, processing') = completed plan processing (installedUnitId p)
+
+
+-- ------------------------------------------------------------
+-- * Executing plans
+-- ------------------------------------------------------------
+
+-- | The set of results we get from executing an install plan.
+--
+type BuildResults failure result = Map UnitId (Either failure result)
```
Comment (ezyang) [inline cabal-install/Distribution/Client/InstallPlan.hs]: NB: This collides with BuildResults defined in Types

**Comment by ezyang** [comment]
Merged version at #3611.

**Comment by 23Skidoo** [comment]
Closing in favour of #3611, which @dcoutts should be able to push to.

**Comment by dcoutts** [comment]
Heh, ok, I was also rebasing. Will compare with #3611.

**Pull request closed without merging**